File: contest/atcoder/abc164/d/main.cpp
#include<iostream>
#include<utility>
#include<string>
#include<vector>
#include<map>

using namespace std;
#define cerr cerr << "[DBG] "
#define DBG(x) cerr << #x << ": " << x << endl
// http://genkisugimoto.com/jp/blog/procon/2015/04/15/print-debug-technique-in-cpp.html
template<typename T1, typename T2> ostream& operator<<(ostream& s, const pair<T1, T2>& p) {return s << "(" << p.first << ", " << p.second << ")";}

template<typename T> ostream& operator<<(ostream& s, const vector<T>& v) { for (int i = 0; i < v.size(); ++i) { s << v[i]; if (i < v.size() - 1) s << "\t"; } return s; }

typedef long long ll;

#define divisor 2019

main(){
    string s;
	cin >> s;
    vector<int> a;
    int n=s.size();
    for(int i=0;i<n;i++){
        a.push_back(s[i]-'0');
    }
    int curr[divisor];
    int prev[divisor];
    for(int i=0;i<divisor;i++){
        curr[i]=prev[i]=0;
    }

    int result = 0;
    for(int i=0;i<n;i++){
        curr[a[i]]++;
        if(i!=0){
            for(int ii=0;ii<divisor;ii++){
                if(prev[ii]==0) continue;
                int rem=(ii*10+a[i])%divisor;
                curr[rem]+=prev[ii];
            }
        }
        result+=curr[0];
        for(int ii=0;ii<divisor;ii++){
            prev[ii]=curr[ii];
            curr[ii]=0;
        }
    }

	cout << result << endl;
}

File: contest/atcoder/abc164/d/main.ts
import * as fs from "fs";
const input = (fs.readFileSync("/dev/stdin", "utf8") as string).split("\n");
const a = input[0].split("").map(x => +x);

const divisor = 2019;

let result = 0;
let prev = new Array<number>(divisor).fill(0);
let curr = new Array<number>(divisor).fill(0);
for (let i = 0; i < a.length; i++) {
	curr[a[i]]++;
	if (i !== 0) {
		for (let ii = 0; ii < divisor; ii++) {
			if (prev[ii] === 0) continue;
			const rem = (ii * 10 + a[i]) % divisor;
			curr[rem] += prev[ii];
		}
	}
	result += curr[0];
	prev = curr;
	curr = new Array<number>(divisor).fill(0);
}
console.error(prev);

/*
const dp = new Array<Array<number>>();
for (let i = 0; i < a.length; i++) {
	dp.push(new Array<number>(divisor).fill(0));
}

for (let i = 0; i < a.length; i++) {
	dp[i][a[i]]++;
	if (i === 0) continue;
	for (let ii = 0; ii < divisor; ii++) {
		if (dp[i - 1][ii] === 0) continue;
		const rem = (ii * 10 + a[i]) % divisor;
		dp[i][rem] += dp[i - 1][ii];
	}
}

let result = 0;
for (let i = 0; i < a.length; i++) {
	result += dp[i][0];
}
*/
console.log(result);
